pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go

```diff
@@ -0,0 +1,333 @@
+	lables := map[string]string{
+		oam.LabelAppName:              o.AppName,
+		oam.LabelAppRevision:          o.AppRevision.Name,
+		oam.LabelAppRevisionHash:      o.AppRevision.Labels[oam.LabelAppRevisionHash],
+		oam.LabelAppComponent:         compName,
+		oam.LabelAppComponentRevision: compRevisionName,
+	}
+	// pass application's all labels to workload/trait
+	return util.MergeMapOverrideWithDst(lables, o.AppLables)
+}
+
+func (o *Options) setWorkloadLables(wl *unstructured.Unstructured, commonLables map[string]string) {
+	// add more workload-specific labels here
+	util.AddLabels(wl, map[string]string{oam.LabelOAMResourceType: oam.ResourceTypeWorkload})
+	util.AddLabels(wl, commonLables)
+
+	/* NOTE a workload has these possible labels
+	   app.oam.dev/app-revision-hash: ce053923e2fb403f
+	   app.oam.dev/appRevision: myapp-v2
+	   app.oam.dev/component: mycomp
+	   app.oam.dev/name: myapp
+	   app.oam.dev/resourceType: WORKLOAD
+	   app.oam.dev/revision: mycomp-v2
+	   workload.oam.dev/type: kube-worker
+	*/
+}
+
+func (o *Options) setTraitLables(trait *unstructured.Unstructured, commonLables map[string]string) {
+	// add more trait-specific labels here
+	util.AddLabels(trait, map[string]string{oam.LabelOAMResourceType: oam.ResourceTypeTrait})
+	util.AddLabels(trait, commonLables)
+
+	/* NOTE a trait has these possible labels
+	   app.oam.dev/app-revision-hash: ce053923e2fb403f
+	   app.oam.dev/appRevision: myapp-v2
+	   app.oam.dev/component: mycomp
+	   app.oam.dev/name: myapp
+	   app.oam.dev/resourceType: TRAIT
+	   app.oam.dev/revision: mycomp-v2
+	   trait.oam.dev/resource: service
+	   trait.oam.dev/type: ingress // already added in render phase
+	*/
+}
+
+// workload and trait both have these annotations
+func (o *Options) setAnnotations(obj *unstructured.Unstructured) {
+	// pass application's all annotations
+	util.AddAnnotations(obj, o.AppAnnotations)
+	// remove useless annotations for workload/trait
+	util.RemoveAnnotations(obj, []string{
+		oam.AnnotationAppRollout,
+		oam.AnnotationRollingComponent,
+		oam.AnnotationInplaceUpgrade,
+	})
+}
+
+func (o *Options) setNamespace(obj *unstructured.Unstructured) {
+	// only set app's namespace when namespace is unspecified
+	// it's by design to set arbitrary namespace in render phase
+	if len(obj.GetNamespace()) == 0 {
+		obj.SetNamespace(o.AppNamespace)
+	}
+}
+
+func (o *Options) setOwnerReference(obj *unstructured.Unstructured) {
+	obj.SetOwnerReferences([]metav1.OwnerReference{*o.AppOwnerRef})
+}
+
+func (o *Options) setWorkloadRefToTrait(wlRef runtimev1alpha1.TypedReference, trait *unstructured.Unstructured) error {
+	traitType := trait.GetLabels()[oam.TraitTypeLabel]
+	traitDef := o.AppRevision.Spec.TraitDefinitions[traitType]
+	workloadRefPath := traitDef.Spec.WorkloadRefPath
+	// only add workload reference to the trait if it asks for it
+	if len(workloadRefPath) != 0 {
+		if err := fieldpath.Pave(trait.UnstructuredContent()).SetValue(workloadRefPath, wlRef); err != nil {
```

Comment:
we should write test case for it, this is a feature which used to rely on AppConfig